org.eclipse.xtext/src/org/eclipse/xtext/generator/trace/node/TracingExtensions.xtend

```diff
@@ -0,0 +1,96 @@
+/*******************************************************************************
+ * Copyright (c) 2017 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.generator.trace.node
+
+import java.util.function.Function
+import org.eclipse.emf.ecore.EFactory
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.emf.ecore.EStructuralFeature
+import org.eclipse.xtend.lib.macro.AbstractClassProcessor
+import org.eclipse.xtend.lib.macro.Active
+import org.eclipse.xtend.lib.macro.TransformationContext
+import org.eclipse.xtend.lib.macro.declaration.InterfaceDeclaration
+import org.eclipse.xtend.lib.macro.declaration.MutableClassDeclaration
+import org.eclipse.xtend.lib.macro.declaration.ResolvedMethod
+import org.eclipse.xtext.generator.trace.ILocationData
+
+/**
+ * @author Sven Efftinge - Initial contribution and API
+ */
+@Active(TracingExtensionsProcessor)
+annotation TracingExtensions {
+	Class<? extends EFactory>[] value
+}
+
+class TracingExtensionsProcessor extends AbstractClassProcessor {
+	
+	override doTransform(MutableClassDeclaration annotatedClass, extension TransformationContext context) {
+		val eobjectType = EObject.newTypeReference()
+		annotatedClass.extendedClass = TracingSugar.newTypeReference()
+		val annotationType = findTypeGlobally(TracingExtensions)
+		val factories = annotatedClass.findAnnotation(annotationType)?.getClassArrayValue("value")
+		if (factories === null) {
+			return;
+		}
+		for (f : factories.map[type].filter(InterfaceDeclaration)) {
+			for (t: f.declaredMethods.filter[simpleName.startsWith('create') && parameters.empty].map[returnType]) {
+				for (getter : t.allResolvedMethods.filter[ isGetter]) {
+					val rt = getter.resolvedReturnType
+					if (allowedLowerCaseTypeNames.contains(rt.type.simpleName.toLowerCase)) {
+						annotatedClass.addMethod(getter.tracerName) [
+							returnType = IGeneratorNode.newTypeReference()
+							addParameter('target', t)
+							body = '''
+								�EStructuralFeature� feature = target.eClass().getEStructuralFeature("�getter.featureName�");
+								�ILocationData� location = this.location(target, feature, -1);
+								�CompositeGeneratorNode� trace = this.trace(location);
+								this.append(trace, target.�getter.declaration.simpleName�());
+								return trace;
+							'''
+						]
+					} else if (eobjectType.isAssignableFrom(rt)) {
```

Comment:
I think it would be nice to have the `getFoo(target, stringProvider)` for any kind of EStructuralFeature-based Getter, even for Strings and floats. They may require some formatting, e.g. turning a string into a getter name or deciding for a float about the number digits after the dot.